From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lokbondo Kung <lokokung@google.com>
Date: Wed, 28 Oct 2026 10:54:04 -0700
Subject: [M156] [security] Make sure that mPipeline in CreatePipelineEvent is
 cleared.

- Note that some of the test changes from the original change is not merged
  back because they aren't necessary since before we passed the error
  directly and did not keep a lingering copy of it.

Original change's description:
> [security] Make sure that mPipeline in CreatePipelineEvent is cleared.
>
> - As per the bug below, the mPipeline reference may live if an error
>   occurs as a part of initializing the pipeline. Because we still hold
>   a reference to the resulting pipeline in the error case, we can run
>   into a teardown race where we reference the Vulkan object when it is
>   no longer valid.
> - This change just ensures that we always move the member variable out
>   after the Complete call while holding the device lock.
> - Update existing tests to verify the fix. Note that the crucial change
>   to the test is to pass Ref<PipelineMock> with std::move so that the
>   last reference is dropped the way it would be in actual production.
>
> Bug: 575150692
> Change-Id: Id1497f9fde96595bdba3b675415d2f07bc709a73
> Reviewed-on: https://dawn-review.googlesource.com/c/dawn/+/352215
> Commit-Queue: Corentin Wallez <cwallez@chromium.org>
> Auto-Submit: Loko Kung <lokokung@google.com>
> Reviewed-by: Corentin Wallez <cwallez@chromium.org>

(cherry picked from commit 2afd6c8a512fed7956c8d2579e10c2209b23960d)

Bug: 576948305,575150692
Change-Id: Id1497f9fde96595bdba3b675415d2f07bc709a73
Reviewed-on: https://dawn-review.googlesource.com/c/dawn/+/353916
Reviewed-by: Corentin Wallez <cwallez@chromium.org>

diff --git a/src/dawn/native/CreatePipelineAsyncEvent.cpp b/src/dawn/native/CreatePipelineAsyncEvent.cpp
index cbfdc5c72e445c2aaa4ab3cd7278580b9823835d..7275e7d6b5b59b4278070c1ae12387c6db6381ad 100644
--- a/src/dawn/native/CreatePipelineAsyncEvent.cpp
+++ b/src/dawn/native/CreatePipelineAsyncEvent.cpp
@@ -204,8 +204,8 @@ void CreatePipelineAsyncEvent<PipelineType, CreatePipelineAsyncCallbackInfo>::Co
                 const std::string& label = mPipeline->GetLabel();
                 mPipeline = PipelineType::MakeError(device.Get(), label.c_str());
             }
-            pipeline = std::move(mPipeline);
         } else if (mError != nullptr) {
+            mPipeline = nullptr;
             switch (mError->GetType()) {
                 case InternalErrorType::Validation:
                     status = WGPUCreatePipelineAsyncStatus_ValidationError;
@@ -217,8 +217,8 @@ void CreatePipelineAsyncEvent<PipelineType, CreatePipelineAsyncCallbackInfo>::Co
             message = mError->GetFormattedMessage();
         } else {
             AddOrGetCachedPipeline();
-            pipeline = std::move(mPipeline);
         }
+        pipeline = std::move(mPipeline);
     }
 
     if (mCallback) {
diff --git a/src/dawn/tests/unittests/native/CreatePipelineAsyncEventTests.cpp b/src/dawn/tests/unittests/native/CreatePipelineAsyncEventTests.cpp
index b3de843931e5329c639f91fff35f1eed2c6e6ea4..776d7da9aab99c1a2a5d727d4f13680161a7d1a5 100644
--- a/src/dawn/tests/unittests/native/CreatePipelineAsyncEventTests.cpp
+++ b/src/dawn/tests/unittests/native/CreatePipelineAsyncEventTests.cpp
@@ -46,9 +46,12 @@ namespace dawn::native {
 namespace {
 
 using ::testing::_;
+using ::testing::ByMove;
 using ::testing::HasSubstr;
+using ::testing::InSequence;
 using ::testing::IsNull;
 using ::testing::MockCppCallback;
+using ::testing::Return;
 
 using MockComputePipelineAsyncCallback = MockCppCallback<
     void (*)(wgpu::CreatePipelineAsyncStatus, wgpu::ComputePipeline, wgpu::StringView)>;
@@ -86,13 +89,16 @@ TEST_F(CreatePipelineAsyncEventTests, InitializationValidationErrorInCreateRende
     desc.depthStencil = &ds;
     Ref<RenderPipelineMock> renderPipelineMock =
         RenderPipelineMock::Create(mDeviceMock, FromCppAPI(&desc));
+    RenderPipelineMock* renderPipelineMockPtr = renderPipelineMock.Get();
 
     ON_CALL(*renderPipelineMock.Get(), InitializeImpl)
         .WillByDefault(testing::Return(testing::ByMove(
             DAWN_MAKE_ERROR(InternalErrorType::Validation, std::string(kErrorMessage)))));
     ON_CALL(*mDeviceMock.get(), CreateUninitializedRenderPipelineImpl)
-        .WillByDefault(testing::Return(testing::ByMove(renderPipelineMock)));
+        .WillByDefault(Return(ByMove(std::move(renderPipelineMock))));
 
+    InSequence s;
+    EXPECT_CALL(*renderPipelineMockPtr, DestroyImpl).Times(1);
     EXPECT_CALL(mockRenderPipelineCb, Call(wgpu::CreatePipelineAsyncStatus::ValidationError,
                                            IsNull(), HasSubstr(kErrorMessage)))
         .Times(1);
@@ -100,8 +106,6 @@ TEST_F(CreatePipelineAsyncEventTests, InitializationValidationErrorInCreateRende
     device.CreateRenderPipelineAsync(&desc, wgpu::CallbackMode::AllowProcessEvents,
                                      mockRenderPipelineCb.Callback());
     ProcessEvents();
-
-    EXPECT_CALL(*renderPipelineMock.Get(), DestroyImpl).Times(1);
 }
 
 // Verify CreateComputePipelineAsync and the internal CreateComputePipelineAsyncEvent behavior
@@ -117,13 +121,16 @@ TEST_F(CreatePipelineAsyncEventTests, InitializationInternalErrorInCreateRenderP
     desc.depthStencil = &ds;
     Ref<RenderPipelineMock> renderPipelineMock =
         RenderPipelineMock::Create(mDeviceMock, FromCppAPI(&desc));
+    RenderPipelineMock* renderPipelineMockPtr = renderPipelineMock.Get();
 
     ON_CALL(*renderPipelineMock.Get(), InitializeImpl)
         .WillByDefault(testing::Return(testing::ByMove(
             DAWN_MAKE_ERROR(InternalErrorType::Internal, std::string(kErrorMessage)))));
     ON_CALL(*mDeviceMock.get(), CreateUninitializedRenderPipelineImpl)
-        .WillByDefault(testing::Return(testing::ByMove(renderPipelineMock)));
+        .WillByDefault(Return(ByMove(std::move(renderPipelineMock))));
 
+    InSequence s;
+    EXPECT_CALL(*renderPipelineMockPtr, DestroyImpl).Times(1);
     EXPECT_CALL(mockRenderPipelineCb, Call(wgpu::CreatePipelineAsyncStatus::InternalError, IsNull(),
                                            HasSubstr(kErrorMessage)))
         .Times(1);
@@ -131,8 +138,6 @@ TEST_F(CreatePipelineAsyncEventTests, InitializationInternalErrorInCreateRenderP
     device.CreateRenderPipelineAsync(&desc, wgpu::CallbackMode::AllowProcessEvents,
                                      mockRenderPipelineCb.Callback());
     ProcessEvents();
-
-    EXPECT_CALL(*renderPipelineMock.Get(), DestroyImpl).Times(1);
 }
 
 // Test that a long async task's execution won't extend to after the device is dropped.
@@ -173,13 +178,16 @@ TEST_F(CreatePipelineAsyncEventTests, InitializationValidationErrorInCreateCompu
     desc.compute.module = utils::CreateShaderModule(device, kComputeShader.data());
     Ref<ComputePipelineMock> computePipelineMock =
         ComputePipelineMock::Create(mDeviceMock, FromCppAPI(&desc));
+    ComputePipelineMock* computePipelineMockPtr = computePipelineMock.Get();
 
     ON_CALL(*computePipelineMock.Get(), InitializeImpl)
         .WillByDefault(testing::Return(testing::ByMove(
             DAWN_MAKE_ERROR(InternalErrorType::Validation, std::string(kErrorMessage)))));
     ON_CALL(*mDeviceMock.get(), CreateUninitializedComputePipelineImpl)
-        .WillByDefault(testing::Return(testing::ByMove(computePipelineMock)));
+        .WillByDefault(Return(ByMove(std::move(computePipelineMock))));
 
+    InSequence s;
+    EXPECT_CALL(*computePipelineMockPtr, DestroyImpl).Times(1);
     EXPECT_CALL(mockComputePipelineCb, Call(wgpu::CreatePipelineAsyncStatus::ValidationError,
                                             IsNull(), HasSubstr(kErrorMessage)))
         .Times(1);
@@ -187,8 +195,6 @@ TEST_F(CreatePipelineAsyncEventTests, InitializationValidationErrorInCreateCompu
     device.CreateComputePipelineAsync(&desc, wgpu::CallbackMode::AllowProcessEvents,
                                       mockComputePipelineCb.Callback());
     ProcessEvents();
-
-    EXPECT_CALL(*computePipelineMock.Get(), DestroyImpl).Times(1);
 }
 
 // Verify CreateComputePipelineAsync and the internal CreateComputePipelineAsyncEvent behavior
@@ -198,13 +204,16 @@ TEST_F(CreatePipelineAsyncEventTests, InitializationInternalErrorInCreateCompute
     desc.compute.module = utils::CreateShaderModule(device, kComputeShader.data());
     Ref<ComputePipelineMock> computePipelineMock =
         ComputePipelineMock::Create(mDeviceMock, FromCppAPI(&desc));
+    ComputePipelineMock* computePipelineMockPtr = computePipelineMock.Get();
 
     ON_CALL(*computePipelineMock.Get(), InitializeImpl)
         .WillByDefault(testing::Return(testing::ByMove(DAWN_MAKE_ERROR(
             dawn::native::InternalErrorType::Internal, std::string(kErrorMessage)))));
     ON_CALL(*mDeviceMock.get(), CreateUninitializedComputePipelineImpl)
-        .WillByDefault(testing::Return(testing::ByMove(computePipelineMock)));
+        .WillByDefault(Return(ByMove(std::move(computePipelineMock))));
 
+    InSequence s;
+    EXPECT_CALL(*computePipelineMockPtr, DestroyImpl).Times(1);
     EXPECT_CALL(mockComputePipelineCb, Call(wgpu::CreatePipelineAsyncStatus::InternalError,
                                             IsNull(), HasSubstr(kErrorMessage)))
         .Times(1);
@@ -212,8 +221,6 @@ TEST_F(CreatePipelineAsyncEventTests, InitializationInternalErrorInCreateCompute
     device.CreateComputePipelineAsync(&desc, wgpu::CallbackMode::AllowProcessEvents,
                                       mockComputePipelineCb.Callback());
     ProcessEvents();
-
-    EXPECT_CALL(*computePipelineMock.Get(), DestroyImpl).Times(1);
 }
 
 // Verify that noexcept callbacks work with CreateComputePipelineAsync.
